doc/adrs/XXX_support-server-side-pagination-for-table.adoc

```diff
@@ -0,0 +1,146 @@
+
+===== Disadvantages
+* Complexity: cursor-based pagination can be more complex to implement.
+* User Experience: It may not provide a straightforward way to jump to a specific page.
+* We cannot manage multiple `rowDescription` instances because we are unable to determine the number of rows returned for each description.
+
+==== Conclusion
+
+Performance considerations led us to choose the *cursor-based* pagination approach.
+The constraint of having only one instance of `rowDescription` is not inherently problematic.
+
+=== Material React Table
+
+To enable server-side pagination, we will set `manualPagination: true` in Material React Table.
+This will allow us to take control of pagination on the backend, ensuring that we can efficiently manage large datasets and accurately display the true total number of rows.
+The pagination must be managed within our own state, and listeners must be added to handle pagination state changes, allowing the backend to be called with the updated pagination data.
+
+[source, typescript]
+----
+const [pagination, setPagination] = useState({
+  pageIndex,
+  pageSize,
+});
+
+useEffect(() => {
+  //do something when the pagination state changes
+}, [pagination.pageIndex, pagination.pageSize]);
+
+const table = useMaterialReactTable({
+  onPaginationChange: setPagination,
+  state: { pagination },
+});
+
+
+----
+
+=== Frontend
+
+The pagination data will be added to the `representationId` of the subscription:
+
+[source, typescript]
+----
+`${tableId}?cursor=${cursor}&pageSize=${pageSize}`
+----
+
+=== GraphQL
+
+* Add pagination data in `Table` type
+
+[source, graphql]
+----
+type Table {
+
+  pagination: Pagination
+}
+
+type Pagination {
+  hasPreviousPage: Boolean!
+  hasNextPage: Boolean!
+  rowCount: Int
+}
+----
+
+NOTE: `rowCount` is optional for performance reasons.
+
+NOTE: the `lines` that is passed to the table already has had the pagination logic applied.
+
+=== Backend
+
+The following variables need to be passed to the line semantic candidat expression:
+* cursor (object retrieve with the cursor param from the front)
+* pageSize
+
+== Customize row per page options
+
+Another issue to address is the ability to define the possible values for the number of rows per page.
+To address this, we will enhance the table description capabilities by adding a list of integers corresponding to the options that will be available on the frontend.
+
+* backend:
+
+[source, java]
+----
+public final class TableDescription {
+    private PaginationOptions paginationOptions;
+}
+
+public final class PaginationOptions {
+    private List<Integer> rowsPerPageOptions;
+}
+----
+
+Note: This customization may also be available at the view DSL level.
+
+* frontend:
+
+[source, typescript]
+----
+const table = useMaterialReactTable({
+  muiPaginationProps: {
+    rowsPerPageOptions: [5, 10, 20],// data from backend
+  },
+});
+----
+
+== Status
+
+Work in progress
+
+== Consequences
```

Comment:
This new version of the ADR is great, it captures the complexity of the problem but I just want to clarify a couple of things that will happen as a consequence of this ADR. It will have a lot of impact on all users of tables and on the development of future features so I want to make it very explicit.

Could you had, as consequences, something a bit like this:

=== New AQL services for the navigation

Support for a cursor-based backend pagination cannot be properly done without new AQL services to navigate in the graph of EObjects. Those [XPath](https://www.w3.org/TR/1999/REC-xpath-19991116/) related services will start their life in Sirius Web but they should ultimately be upstreamed in the Acceleo project in `org.eclipse.acceleo.query.services.XPathServices` directly. Among the services we will need, we will have:

- `EObject#following(int size)`: Used to retrieve the next `size` objects following the given one
- `EObject#hasFollowing()`
- `EObject#preceding(int size)`: Used to retrieve the previous `size` objects preceding the given one
- `EObject#hasPreceding()`
- `EObject#collecting(String direction, int size)`: Used to call either `following` or `preceding` depending on the value of `direction` which could either be `backward` or `forward`.


=== Support for paginated data

We will need to retrieve pagination data on top of the raw data, for that the return type of the `semanticCandidatesExpression` will be a `PaginatedData` not unlike Spring Data's `Page`.

```
record PaginatedData<T> (
  List<T> content,
  boolean hasNextPage,
  boolean hasPreviousPage,
  int totalElementCount
)
```

Contrary to what I told you on the other ADR for the view DSL, it will not be possible for us to apply our own filters like `DomainType` or `preconditionExpression` after a paginated `semanticCandidatesExpression`. We will have to provide a service to let specifiers create paginated data from a collection themselves:

- `Collection#toPaginatedData(boolean hasNextPage, boolean hasPreviousPage)`
- `Collection#toPaginatedData(boolean hasNextPage, boolean hasPreviousPage, int totalElementCount)`

Those services and data structure should not be specific to tables and probably be in core parts of Sirius Components.

=== Support for filter, sort, etc

Additional features that will impact the content or the order of the elements will have to be given to specifiers as a variable in the `semanticCandidatesExpression`. In order to provide some basic support for those variables, we will have to provide better services to consider them, probably at least something like this:

- `EObject#collecting(String direction, Filter filter, Sort sort, int size)`

With potentially something like this to sort the results:

```
record Sort(List<SortEntry> entries)

record SortEntry(String featureName, String sortOrder)
```

To be able to indicate that we want to sort a collection of individual by "Family name ascending", then "First name ascending", then "age descending". The final API to sort / filter may not look like this but it will be critical to provide all the relevant variables in the `semanticCandidatesExpression` and maintain XPath inspired AQL services to use the feature. With all that in mind, specifiers planning to use our support for table should be prepared to write something like this as a `semanticCandidatesExpression`:

```
aql:cursor.collecting(direction, filter, sort, size)->toPaginatedData(cursor.hasPreceding(), cursor.hasFollowing())
```

None of those parameters will be ignorable to return a valid result.